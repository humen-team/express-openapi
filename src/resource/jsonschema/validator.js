import { isPlainObject, mapValues } from 'lodash';
import { Validator } from 'jsonschema';

import Reference from '../reference';
import { userIdFormat, uuidFormat } from './formats';

/* Replace `Reference` instances in the schema with an concrete path.
 *
 * OpenAPI 2.0 and 3.0.0 expect different root paths for resources. We avoid committing
 * to one path or another by using a `Reference` object instance and deferring spec building
 * until as late as possible.
 *
 * However, internal references may be needed for resource validation, so we replace
 * `Reference` instances with a _generic_ path during validation.
 */
function dereference(object, validator, path = '') {
    // recurse into lists
    if (Array.isArray(object)) {
        return object.map(
            (item) => dereference(item, validator, `${path}[]`),
        );
    }

    // recurse into objects
    if (isPlainObject(object)) {
        return mapValues(
            object,
            (value, key) => {
                if (key !== '$ref') {
                    // not a ref
                    return dereference(value, validator, `${path}.${key}`);
                }
                if (!(value instanceof Reference)) {
                    // not a valid ref
                    throw new Error('Expected $ref to be instance of Reference; did you use Resource.toRef()?');
                }
                // use '/foo' as the generic reference path
                const refId = value.id;

                // recurse to the referenced schema
                const { id, ...referenceSchema } = dereference(value.resource.schema, validator);

                // save the reference schema under the reference id *locally* within this validator.
                validator.addSchema({ id: refId, ...referenceSchema });

                // return the generic reference path in place of the Reference
                return refId;
            },
        );
    }

    // base case
    return object;
}

/* Create a JSON Schema validator for a specific resource.
 */
export default function createValidator(resource) {
    const validator = new Validator();

    validator.customFormats.uuid = uuidFormat;
    validator.customFormats['user-id'] = userIdFormat;

    const schema = dereference(resource.schema, validator);
    return {
        validate: (object) => validator.validate(object, schema),
    };
}
