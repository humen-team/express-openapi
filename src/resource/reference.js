import pickVersion from '../versions';

/* A reference to a resource.
 *
 * Enables deferal of $ref encoding as late as possible to account for differences
 * between OpenAPI 2.0 and 3.0.0.
 *
 * At the moment, references are required to form a DAG. Cycles will break.
 */
export default class Reference {
    constructor(resource) {
        this.resource = resource;
    }

    get id() {
        return this.resource.id;
    }

    build(openapiVersion) {
        return pickVersion(this, 'build', openapiVersion)();
    }

    build20() {
        return `#/definitions/${this.id}`;
    }

    build300() {
        return `#/components/schemas/${this.id}`;
    }
}
