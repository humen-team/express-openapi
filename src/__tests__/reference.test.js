import request from 'supertest';

import { Namespace } from '..';
import { newApp } from './app';
import { Parent } from './reference';
import validator, { openapi2, openapi3 } from './validate';

function search() {
    return {
        count: 3,
        items: [
            {
                name: 'parentWithChildAndGrandchild',
                child: {
                    name: 'childWithGrandchild',
                    grandchild: {
                        name: 'grandchild',
                    },
                },
            },
            // TODO: nullable references are non trivial to represent
            /*
            {
                name: 'parentWithChild',
                child: {
                    name: 'child',
                    grandchild: null,
                },
            },
            {
                name: 'parent',
                child: null,
            },
            */
        ],
    };
}

describe('schema references', () => {
    const operation = new Namespace('Parent')
        .search({ output: Parent.toList(), route: search });

    const app = newApp({ operations: [operation] });

    describe('search', () => {
        it('returns a list of parents with references to children', async () => {
            const response = await request(app).get('/parent');

            expect(response.statusCode).toEqual(200);
            expect(response.body.items.length).toEqual(1);
            expect(response.body).toMatchSnapshot();
        });
    });

    describe('openapi 2.0', () => {
        it('is valid', async () => {
            const response = await request(app).get('/openapi/2.0');
            expect(response.statusCode).toEqual(200);
            validator.validate(response.body, openapi2, { throwError: true });
            expect(response.body).toMatchSnapshot();
        });
    });

    describe('openapi 3.0.0', () => {
        it('is valid', async () => {
            const response = await request(app).get('/openapi/3.0.0');
            expect(response.statusCode).toEqual(200);
            validator.validate(response.body, openapi3, { throwError: true });
            expect(response.body).toMatchSnapshot();
        });
    });
});
