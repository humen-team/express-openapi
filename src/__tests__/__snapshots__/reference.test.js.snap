// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema references openapi 2.0 is valid 1`] = `
Object {
  "basePath": "/",
  "definitions": Object {
    "Error": Object {
      "properties": Object {
        "message": Object {
          "type": "string",
          "x-nullable": false,
        },
        "status": Object {
          "type": "integer",
          "x-nullable": false,
        },
      },
      "type": "object",
    },
    "ParentList": Object {
      "properties": Object {
        "count": Object {
          "type": "integer",
          "x-nullable": false,
        },
        "items": Object {
          "items": Object {
            "$ref": "#/definitions/Parent",
          },
          "type": "array",
          "x-nullable": false,
        },
        "limit": Object {
          "type": "integer",
          "x-nullable": false,
        },
        "offset": Object {
          "type": "integer",
          "x-nullable": false,
        },
      },
      "required": Array [
        "items",
        "count",
      ],
      "type": "object",
    },
  },
  "info": Object {
    "description": "REST API",
    "title": "REST API",
    "version": "0.1.0",
  },
  "paths": Object {
    "/parent": Object {
      "get": Object {
        "operationId": "search",
        "parameters": Array [],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "200": Object {
            "description": "ParentList as JSON",
            "schema": Object {
              "$ref": "#/definitions/ParentList",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "Parent",
        ],
      },
    },
  },
  "schemes": Array [
    "http",
  ],
  "swagger": "2.0",
}
`;

exports[`schema references openapi 3.0.0 is valid 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "Error": Object {
        "properties": Object {
          "message": Object {
            "nullable": false,
            "type": "string",
          },
          "status": Object {
            "nullable": false,
            "type": "integer",
          },
        },
        "type": "object",
      },
      "ParentList": Object {
        "properties": Object {
          "count": Object {
            "nullable": false,
            "type": "integer",
          },
          "items": Object {
            "items": Object {
              "$ref": "#/components/schemas/Parent",
            },
            "nullable": false,
            "type": "array",
          },
          "limit": Object {
            "nullable": false,
            "type": "integer",
          },
          "offset": Object {
            "nullable": false,
            "type": "integer",
          },
        },
        "required": Array [
          "items",
          "count",
        ],
        "type": "object",
      },
    },
  },
  "info": Object {
    "description": "REST API",
    "title": "REST API",
    "version": "0.1.0",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/parent": Object {
      "get": Object {
        "operationId": "search",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/ParentList",
                },
              },
            },
            "description": "ParentList as JSON",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "Parent",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "/",
    },
  ],
}
`;

exports[`schema references search returns a list of parents with references to children 1`] = `
Object {
  "count": 3,
  "items": Array [
    Object {
      "child": Object {
        "grandchild": Object {
          "name": "grandchild",
        },
        "name": "childWithGrandchild",
      },
      "name": "parentWithChildAndGrandchild",
    },
    Object {
      "child": Object {
        "grandchild": null,
        "name": "child",
      },
      "name": "parentWithChild",
    },
    Object {
      "child": null,
      "name": "parent",
    },
  ],
}
`;
